js/debouncing.md

```diff
@@ -0,0 +1,78 @@
+Often situations arise, when we want a certain event to fire off only after a certain amount of time has passed. The most common example would be the search bars we see on e-commerce websites like Amazon or Flipkart. In these search bars, when you type something, an API call is made, and results similar to your query are displayed on the screen.
+
+Now imagine making an API call every time a key is pressed on your keyboard. Doesn't sound viable right? Firing an event continuously (in this case, the call to the API) would hamper the performance of the website.
+
+This can be avoided by using the concept of debouncing. Debouncing limits the rate at which a function fires off. In the example above, it drastically reduces the number of API calls that are made to the server.
+
+The setTimeout function of Javascript plays a vital role in implementing debouncing. For those who don't know what setTimeout does, it is a scheduling function provided by Javascript which allows us to run a task after a certain interval of time.
+
+#### Let's see how we can implement debouncing with a simple example,
+
+The very first thing we'll do here is - create a simple input in our HTML file. On this input box, we will call a method `searchData()` for the `onkeyup` event. So that every time the user releases the key after pressing it, `searchData()` is called.
+
+```
+<body>
+    <label for="search">Search</label><br /><br />
+    <input type="text" onkeyup="searchData()" />
+
+    <script src="script.js"></script>
+  </body>
+```
+
+In our Javascript file, we'll define the `searchData()` function and create a counter to track the number of times it gets called.
+
+```
+let counter = 1; // To track number of calls
+
+const searchData = () => {
+  // Let's say this function calls some API and gets us some data
+  console.log("Looking for data..", counter++);
+};
+
+```
+
+#### After running this program, we see something like this -
+
+![first.gif](https://cdn.hashnode.com/res/hashnode/image/upload/v1592509132273/zPQonjL7A.gif)
+
+Notice how for every letter of the word "Michael", the function gets called. Imagine firing off API calls for a bigger query on a much bigger site. That would be very browser intensive and may hinder the performance. To reduce these continuous function calls, we'll make use of the debounce function :
+
+```
+let counter = 1; // To track number of calls
+
+const searchData = () => {
+  // Let's say this function calls some API and gets us some data
+  console.log("Looking for data..", counter++);
+};
+
+// Debounce function
```

Comment:
After the code add a section giving overview of what we are doing. Then explain the code with a dry run example like first user enters M then callDebounce is called then this happens then that and so on. Showing this for first character, second character and last character will make it very clear what's happening.

Reply:
Hey Aditya, I made a few changes. Please look if it looks good now.

Reply:
approved